crates/core/src/error.rs

```diff
@@ -1,810 +1,451 @@
+            backtrace: Some(Backtrace::capture()),
+        })
+    }
+
+    /// Checks if the error has a specific error code.
+    pub fn is_code(&self, code: &str) -> bool {
+        self.error_code
+            .as_ref()
+            .map_or(false, |c| c.as_str() == code)
+    }
+
+    /// Checks if the error is due to an incorrect password
+    pub fn is_incorrect_password(&self) -> bool {
+        self.is_code("C002")
+    }
+
+    /// Creates a new error from a given error.
+    pub fn from<T: std::error::Error + Display + Send + Sync + 'static>(
+        error: T,
+        kind: ErrorKind,
+    ) -> Box<Self> {
+        Box::new(Self {
+            kind,
+            guidance: error.to_string().into(),
+            context: Cow::default(),
+            source: Some(Box::new(error)),
+            error_code: None,
+            docs_url: None,
+            new_issue_url: None,
+            existing_issue_url: None,
+            severity: None,
+            status: None,
+            // `Backtrace::capture()` will check if backtrace has been enabled
+            // internally. It's zero cost if backtrace is disabled.
+            backtrace: Some(Backtrace::capture()),
+        })
+    }
 }
 
-trait RusticErrorMarker: Error {}
-
-impl RusticErrorMarker for CryptoErrorKind {}
-impl RusticErrorMarker for PolynomialErrorKind {}
-impl RusticErrorMarker for IdErrorKind {}
-impl RusticErrorMarker for RepositoryErrorKind {}
-impl RusticErrorMarker for IndexErrorKind {}
-impl RusticErrorMarker for BackendAccessErrorKind {}
-impl RusticErrorMarker for ConfigFileErrorKind {}
-impl RusticErrorMarker for KeyFileErrorKind {}
-impl RusticErrorMarker for PackFileErrorKind {}
-impl RusticErrorMarker for SnapshotFileErrorKind {}
-impl RusticErrorMarker for PackerErrorKind {}
-impl RusticErrorMarker for FileErrorKind {}
-impl RusticErrorMarker for TreeErrorKind {}
-impl RusticErrorMarker for CacheBackendErrorKind {}
-impl RusticErrorMarker for CryptBackendErrorKind {}
-impl RusticErrorMarker for IgnoreErrorKind {}
-impl RusticErrorMarker for LocalDestinationErrorKind {}
-impl RusticErrorMarker for NodeErrorKind {}
-impl RusticErrorMarker for StdInErrorKind {}
-impl RusticErrorMarker for ArchiverErrorKind {}
-impl RusticErrorMarker for CommandErrorKind {}
-impl RusticErrorMarker for VfsErrorKind {}
-impl RusticErrorMarker for std::io::Error {}
-
-impl<E> From<E> for RusticError
-where
-    E: RusticErrorMarker,
-    RusticErrorKind: From<E>,
-{
-    fn from(value: E) -> Self {
-        Self(RusticErrorKind::from(value))
+// Setters for anything we do want to expose publicly.
+//
+// These were initially generated by `derive_setters`,
+// and then manually adjusted to return `Box<Self>` instead of `Self` which
+// unfortunately is not possible with the current version of the `derive_setters`.
+//
+// BEWARE! `attach_context` is manually implemented to allow for multiple contexts
+// to be added and is not generated by `derive_setters`.
```

Comment:
I don't see `derive_setters` (I didn't know about that derive), where is it used?

Reply:
Actually in a lot of places, it replaces the builder pattern for us, in a way.

See `rg 'use derive_setters::'`